source/common/json/json_streamer.cc

```diff
@@ -0,0 +1,149 @@
+#include "source/common/json/json_streamer.h"
+
+#include "source/common/json/json_sanitizer.h"
+
+#include "absl/strings/str_format.h"
+
+namespace Envoy {
+namespace Json {
+
+#ifdef NDEBUG
+#define ASSERT_THIS_IS_TOP_LEVEL                                                                   \
+  do {                                                                                             \
+  } while (0)
+#else
+#define ASSERT_THIS_IS_TOP_LEVEL ASSERT(streamer_.topLevel() == this)
+#endif
```

Comment:
No: commenting here and adding to the code:
```
// To ensure the streamer is being used correctly, we use assertions to enforce
// that only the topmost map/array in the stack is being written to. To make
// this easier to do from the Level classes, we provider Streamer::topLevel() as
// a member function, but this is only needed when compiled for debug.
//
// We only compile Streamer::topLevel in debug to avoid having it be a coverage
// gap. However, assertions fail to compile in release mode if they reference
// non-existant functions or member variables, so we only compile the assertions
// in debug mode.
```

Reply:
Ah, I thought ASSERT compiled away to nothing in NDEBUG but I see it turns into a log message of some sort. Yuck.